yt/data_objects/time_series.py

```diff
@@ -435,6 +440,179 @@ def particle_trajectories(
+            within the tolerance value. If None, simply return the
+            nearest dataset.
+            Default: None.
+        side : str
+            The side of the value to return. Can be 'nearest', 'smaller' or 'larger'.
+            Default: 'nearest'.
+        """
+
+        if side not in ("nearest", "smaller", "larger"):
+            raise ValueError(
+                f"side must be 'nearest', 'smaller' or 'larger', got {side}"
+            )
+
+        # Use a binary search to find the closest value
+        iL = 0
+        iH = len(self._pre_outputs) - 1
+
+        if iL == iH:
+            ds = self[0]
+            if (
+                tolerance is not None
+                and abs(getattr(ds, attribute) - value) > tolerance
+            ):
+                raise ValueError(
+                    f"No dataset found with {attribute} within {tolerance} of {value}."
+                )
+            return ds
+
+        # Check signedness
+        dsL = self[iL]
+        dsH = self[iH]
+        vL = getattr(dsL, attribute)
+        vH = getattr(dsH, attribute)
+
+        if vL < vH:
+            sign = 1
+        elif vL > vH:
+            sign = -1
+        else:
+            raise ValueError(
+                f"{dsL} and {dsH} have both {attribute}={vL}, cannot perform search. "
+                "Try with another key."
+            )
+
+        if isinstance(value, tuple):
+            value = dsL.quan(*value)
+        if isinstance(tolerance, tuple):
+            tolerance = dsL.quan(*tolerance)
+
+        # Short-circuit if value is out-of-range
+        if not (vL * sign < value * sign < vH * sign):
+            iL = iH = 0
+
+        while iH - iL > 1:
+            iM = (iH + iL) // 2
+            dsM = self[iM]
+            vM = getattr(dsM, attribute)
+            if sign * value < sign * vM:
+                iH = iM
+                dsH = dsM
+            elif sign * value > sign * vM:
+                iL = iM
+                dsL = dsM
+            else:  # Exact match
+                dsL = dsH = dsM
+                break
+
+        if side == "smaller":
+            ds_best = dsL if sign > 0 else dsH
+        elif side == "larger":
+            ds_best = dsH if sign > 0 else dsL
+        elif abs(value - getattr(dsL, attribute)) < abs(
+            value - getattr(dsH, attribute)
+        ):
+            ds_best = dsL
+        else:
+            ds_best = dsH
+
+        if tolerance is not None:
+            if abs(value - getattr(ds_best, attribute)) > tolerance:
+                raise ValueError(
+                    f"No dataset found with {attribute} within {tolerance} of {value}."
+                )
+        return ds_best
+
+    def get_by_time(
+        self,
+        time: Union[unyt_quantity, tuple],
+        tolerance: Union[None, unyt_quantity, tuple] = None,
+        side: Union[
+            Literal["nearest"], Literal["smaller"], Literal["larger"]
+        ] = "nearest",
```

Comment:
Unions of `Literal` are actually simpler to express as

```suggestion
        side: Literal["nearest", "smaller", "larger"] = "nearest",
```